lib/ecto/schema.ex

```diff
@@ -1329,17 +1329,22 @@ defmodule Ecto.Schema do
     end
   end
 
-  defp do_load(struct, types, map, loader) when is_map(map) do
+  def __load__(schema, prefix, source, context, data, loader) when is_map(data) do
+    __load__(schema, prefix, source, context, Map.to_list(data), loader)
+  end
+
+  @doc false
+  def do_load(struct, types, data, loader) when is_list(data) do
     Enum.reduce(types, struct, fn
       {field, type}, acc ->
-        case Map.fetch(map, Atom.to_string(field)) do
-          {:ok, value} -> Map.put(acc, field, load!(struct, field, type, value, loader))
-          :error -> acc
+        case Enum.find(data, fn {key, _value} -> key == field or key == Atom.to_string(field) end) do
```

Comment:
needed to look for both atom as well as string keys. If it was just atom keys (`Keyword.fetch`), some integration & association tests were failing.